### ISSUE TITLE
R package GPU support

### ISSUE THREAD

**Issue by RAMitchell** [dmlc/xgboost#2625]
This post is to further explain the call for contribution in #2305 at the request of @SteveBronder.

We currently have support in the GPU plug-in for the Python package and the CLI interface. The reason I was able to easily integrate the GPU plug-in into the Python package is because the installation of the package only requires a prebuilt library file (xgboost.dll/libxgboost.so). The Python package is completely independent of how the XGBoost binary is actually built. 

The R package appears to have a completely different way of building the XGBoost binary from the conventional make or cmake build system. This is important because the GPU version of XGBoost needs to be built in a specific way using the NVCC compiler. Additionally when building the R package on Windows it seems to require the MinGW compiler - cuda code cannot be built using MinGW, only MSVC when using the Windows platform.

So I see the solution to this as finding a way to build the R package using a prebuilt library file.

I have almost no experience with R so feel free to correct me, ask questions or suggest alternatives.

**Comment by khotilov** [comment]
I've previously done a proof-of-concept hack to run gpu-accelerated xgboost in R, in a way that was kind of similar to mxnet R-package. I just need to dig out from my back log...

**Comment by SteveBronder** [comment]
Thanks for posting this!

> Additionally when building the R package on Windows it seems to require the MinGW compiler - cuda code cannot be built using MinGW, only MSVC when using the Windows platform.

I don't see anywhere in the R for windows [faq](https://cran.r-project.org/bin/windows/base/old/2.0.0/rw-FAQ.html) that states you need to use MinGW. I believe the user can simply set the default compiler to Cygwin, msvc, etc.

> So I see the solution to this as finding a way to build the R package using a prebuilt library file.

CRAN does not allow package submissions that contain binary executable code. Not sure if a .so file falls under that category.

> Source packages may not contain any form of binary executable code

[CRAN policies](https://cran.r-project.org/web/packages/policies.html)

@khotilov very excited to see your implementation!

I can't be of much help atm, but I should have time in the next couple weeks to lend a hand

**Comment by jeroen** [comment]
R and all packages are compiled with the [rtools](https://cran.r-project.org/bin/windows/Rtools/) toolchain which currently is a build of `mingw-w64` with `gcc-4.9.3`. Linking to binaries built with other compiles is often difficult. 

You may not ship binaries in the source pkg, but you could create a script or function that lets the user download required binaries, either manually or automatically during installation.

**Comment by khotilov** [comment]
@jeroen As with mxnet, the plan would be to have some special build instructions (dynamically linking an xgboost library) for advanced users, plus some pre-built binary packages in a drat repo. An idea of downloading binaries either during installation or manually is interesting and I'll think more about it. My current sense though is that such approach could be hard to maintain.

**Comment by Laurae2** [comment]
ping @guolinke if you are interested into helping making xgboost compile with Visual Studio + optional NVCC, with automatic fallback to MinGW like we did for https://github.com/Microsoft/LightGBM/pull/584

**Comment by guolinke** [comment]
@Laurae2 
The key problem is hacking of R object. Thus we can avoid including R`s header file, and compile the unify dll/so for both python/R package. 

refer to https://github.com/Microsoft/LightGBM/blob/master/include/LightGBM/R_object_helper.h 
and https://github.com/Microsoft/LightGBM/blob/master/R-package/src/install.libs.R

**Comment by tqchen** [comment]
Also refer to mxnet's approach. https://github.com/apache/incubator-mxnet/tree/master/R-package

The general idea is to build NVCC related libraries via MSVC, while build R wrapper part with Rtools and  dynamically linked the C API. The API design is set up to do this. Except that under that mode we cannot distribute the windows related installations.

**Comment by khotilov** [comment]
One hurdle in the two dynamically linked libraries solution is that we have some R-backed customizations in https://github.com/dmlc/xgboost/blob/master/R-package/src/xgboost_custom.cc
In my previous POC hack, I've simply disregarded all these customizations, loosing the ability to see any internal LOG messages or to use the standard set.seed() when running from R. I still don't see any elegant way to decouple those without making too many changes.

Last weekend, I've tried a single-library approach, building the whole R-package shared library with cmake, and linking it to the R shared library. That seems to work, at least on linux and windows. In windows though, as Boromir used to say, one does not simply link R.dll with MSVC. It requires jumping through some hoops to create an R.lib first. I'm an amateur with cmake, so I would need some more time to cleanly organize the procedure.

@guolinke I'm curious what specific licensing issues forced you to go the wrapper route for decouple from R? Is it a problem to link anything from R when using MSVC?

**Comment by guolinke** [comment]
@khotilov 
R's source code is under GPL license, so we cannot include it to our repo with MIT license. 
The wrapper only contains some routes to access R's objects.

**Comment by khotilov** [comment]
I've just merged #2732 . I think it's ready for use.

@guolinke Is it a Microsoft-specific restriction? On CRAN, there are many compiled packages with MIT or Apache license, and my understanding is that those licenses are considered as compatible with GPL-3.

**Comment by rknimmakayala** [comment]
Appreciate if anyone can share the list of steps for installing xgboost on R for GPUS and my os is windows7.

**Comment by khotilov** [comment]
@rknimmakayala http://xgboost.readthedocs.io/en/latest/build.html#installing-r-package-with-gpu-support

I'm closing this issue since the GPU-enabled R package build procedure works, as far I've seen and heard. For a discussion of the distribution of binary GPU builds it would make sense to have another issue.
